### PR TITLE
Fix Iceberg lineage for windows 

### PR DESCRIPTION
**Pull request by sezruby**
### What is the context for this pull request?

 - **Tracking Issue**: n/a
 - **Parent Issue**: n/a
 - **Dependencies**: n/a
 
### What changes were proposed in this pull request?

Fix for inconsistent result of Iceberg input_file_name() between Windows & Linux

### Does this PR introduce _any_ user-facing change?

Yes, fix iceberg lineage related bug on Windows

### How was this patch tested?

tested on local

### CODE REVIEW

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala [hunk 1]
```diff
@@ -123,14 +125,26 @@ class IcebergRelation(spark: SparkSession, override val plan: DataSourceV2Relati
...
+    //   original file path: file:///path/to/file or file:/path/to/file
+    //   input_file_name(): /path/to/file
+    if (System.getProperty("os.name").toLowerCase(Locale.ROOT).contains("windows")) {
+      fileIdTracker.getFileToIdMap.toSeq.map { kv =>
+        (kv._1._1.stripPrefix("file:/"), kv._2)
+      }
+    } else {
+      fileIdTracker.getFileToIdMap.toSeq.map { kv =>
+        (kv._1._1.replaceAll("^file:/{1,3}", "/"), kv._2)
+      }
```
Comment (andrei-ionescu) [inline src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala:147]:
```suggestion
    val isWindowsOS =
      System.getProperty("os.name").toLowerCase(Locale.ROOT).contains("windows")
    val replacement = if (isWindowsOS) { "" } else { "/" }
    fileIdTracker.getFileToIdMap.toSeq.map { kv =>
      (kv._1._1.replaceAll("^file:/{1,3}", replacement), kv._2)
```

**Review by andrei-ionescu — APPROVED** [review]
Approved plus a suggestion

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala [hunk 2]
```diff
@@ -123,14 +125,26 @@ class IcebergRelation(spark: SparkSession, override val plan: DataSourceV2Relati
    *
    * File paths should be the same format as "input_file_name()" of the given relation type.
    * For [[IcebergRelation]], each file path should be in this format:
-   *   `/path/to/file`
+   *   `/path/to/file` or `X:/path/to/file`
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala:128]: `X:` is only for Windows filesystem (e.g., `file:/`), but not applicable for remote storage (e.g., `abfs://`)? If so, could you clarify this a bit in the comment? For example, we can say that we need to take care of a special care if the path URI's scheme is `file://` depending on the OS.

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala [hunk 3]
```diff
@@ -123,14 +125,26 @@ class IcebergRelation(spark: SparkSession, override val plan: DataSourceV2Relati
...
   override def lineagePairs(fileIdTracker: FileIdTracker): Seq[(String, Long)] = {
-    fileIdTracker.getFileToIdMap.toSeq.map { kv =>
-      (kv._1._1.replaceAll("^file:/{1,3}", "/"), kv._2)
+    // For Windows,
+    //   original file path: file:/C:/path/to/file
+    //   input_file_name(): C:/path/to/file
+    // For Linux,
+    //   original file path: file:///path/to/file or file:/path/to/file
+    //   input_file_name(): /path/to/file
+    if (System.getProperty("os.name").toLowerCase(Locale.ROOT).contains("windows")) {
```
Comment (imback82) [inline src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala:140]: Does `Path.WINDOWS` work?

**Review by imback82 — COMMENTED** [review]
LGTM (few minor comments) and I also verified on Windows. Thanks for the fix, @sezruby!

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala (hunk 1 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala]: I prefer the current one; I think it looks clearer for os specific impl.
Thanks!

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala (hunk 2 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala]: Added some comment.

File: src/test/scala/com/microsoft/hyperspace/index/HyperspaceSuite.scala [hunk 4]
```diff
@@ -111,7 +111,8 @@ trait HyperspaceSuite extends SparkFunSuite with SparkInvolvedSuite {
     // "inheriting conflicting members" issue.
     val path = Utils.createTempDir()
     path.delete()
-    try f(path.toString)
+    val pathStr = new Path(path.toString).toString
```
[the remark this answers is not in the document]
Comment (imback82) [reply src/test/scala/com/microsoft/hyperspace/index/HyperspaceSuite.scala:115]: Could you put the actual example in the comment (and please mention the Iceberg issue/example as well)? Just looking at the code, it's not easy to see how this creates "environment specific" path.

**Review by imback82 — APPROVED** [review]

**Review by imback82 — APPROVED** [review]

File: src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala (hunk 1 above)
Comment (andrei-ionescu) [reply src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala]: I understand your preference... and in the same time I would contradict you saying " it looks clearer for os specific impl" - I don't think there can be anything clearer than `if (isWindowsOS)` construct that I proposed :).

Comment (imback82) [reply src/main/scala/com/microsoft/hyperspace/index/sources/iceberg/IcebergRelation.scala]: either approach seems fine to me.

**Pull request merged**